### ISSUE TITLE
Confused with `jax.vmap` in `FourierPositionEncoding`

### ISSUE THREAD

**Issue by ShoufaChen** [deepmind/deepmind-research#269]
https://github.com/deepmind/deepmind-research/blob/30799687edb1abca4953aec507be87ebe63e432d/perceiver/position_encoding.py#L153-L174

Hi, thanks for your great work.

I was confused with the `jax.vmap` usage in the above position.

From my understanding, `jax.vmap` here is automatically generating FourierPositionEncoding for a **batch**.

Is that right?

If yes, I don't know why use `jax.vmap` here because all samples in a batch share the same FourierPositionEncoding.
In other words, I think it is ok to create the FourierPositionEncoding for a single instance and broadcast it to meet the batch dimension.

Please let me know if I have a misunderstanding.

Thanks in advance.

**Comment by fding** [comment]
Hi @ShoufaChen ,

You're right that often, samples in the batch have the same Fourier position encodings, because we want to compute position encodings for all points in the input image. However, in general, we don't have to compute position encodings for all points in the input image. For example, in video auto-encoding, we subsample points to decode at, and compute position encodings only for those points. The vmap here allows us to compute position encodings for a batch where the positions are not necessarily the same across the batch.

Hopefully this answers your question. Let us know if you have any further questions!

**Comment by ShoufaChen** [comment]
Hi @fding ,

Thanks for your clarification.

Congratulations on your awesome work again.
